### PR TITLE
#35 Extend GLSPServerContribution to support embedded server launch

### PR DESCRIPTION
**Pull request by tortmayr**
Base framework:
- Refactor & document GLSPServerContribution API
-- Rename "start" method to connect.
-- Introduce optional launch method, which can be implemented in embedded server scenarios
-- Add isServerRunning() method, which checks if currently a server instance is running
-Introduce `GLSPServerLaunchOptions`.
- Introduce  a common base implementation for connecting to Java-based GLSP server via socket  ('JavaSocketGLSPServerContribution')

Example:
- Add 'download:exampleServer' script to root package.json
- Refactor WorkflowGLSPServerContribution to support embedded launching of the glsp server
- Add/rework tasks

The example can now be launched in default mode, which will be using a embedded GLSP server and in DEBUG mode which will not launch a server, and expects that the server is also running at the given port. (Previous default behavior)

Fixes eclipse-glsp/glsp/issues/35

### CODE REVIEW

**Comment by tortmayr** [comment]
(Readme will be updated as part of eclipse-glsp/issues/136

Short instructions on how to run/test it:

Per default the GLSP Workflow Theia example will now launch the glsp server as well (embedded).
The root package.json has a script, that automatically downloads the latest sever snapshot version from maven. 
`yarn download:exampleServer`
This script is currently not included in the build (as it prolongs the build time) and has to be executed manually. 

Then the example can be started as before (either via corresponding vscode task or via  yarn start in the browser-app directory.

If you wan't to use an external server (e.g. started from the Eclipse IDE for debugging purposes) you can use the corresponding debug vscode task. (or use `yarn start:debug in the terminal)

**Comment by tortmayr** [comment]
@martin-fleck-at  Feel free to review, but please do not merge this. I have to verify if we need a CQ for the `mvn-artifact-download` dependency. Until then merging is blocked.

**Comment by tortmayr** [comment]
No need for a CQ. We can merge after successful review

File: packages/theia-integration/src/node/java-socket-glsp-server.ts
Comment (martin-fleck-at) [inline packages/theia-integration/src/node/java-socket-glsp-server.ts:81]: It seems like this method is only calls once the contribution starts (at least for the workflow case where we bind the contribution as a singleton), so I am not sure if the evaluation of the multiClient parameter ever is relevant because initially is the server is never running. It could be I am missing something so maybe you can explain that a bit ;-)

**Review by martin-fleck-at — COMMENTED** [review]
Looks good to me in general, thank you very much, Tobias! I found some minor issues in the formatting but they do not have to prevent merging this change. I also have some question about the 'multiClient' case where you can hopefully enlighten me ;-)

File: packages/theia-integration/src/node/java-socket-glsp-server.ts
Comment (tortmayr) [reply packages/theia-integration/src/node/java-socket-glsp-server.ts:81]: You are absolutely right. The evaluation of the multiClient parameter is redundant here. I will remove it.

**Review by martin-fleck-at — APPROVED** [review]
Changes look great to me now, thank you for taking the time to address everything!

**Pull request merged**
